src/tasks/gitea/gitea-drone-oauth.ts

```diff
@@ -0,0 +1,224 @@
+  try {
+    const response = await apiClient.readNamespacedSecret(k8sSecretName, k8sNamespace)
+    const {
+      body: { data },
+    } = response
+    return {
+      clientId: data.clientId,
+      clientSecret: data.clientSecret,
+    } as DroneSecret
+  } catch (e) {
+    return undefined
+  }
+}
+export async function getGiteaAuthorizationHeaderCookies(
+  giteaUrl: string,
+  droneLoginUrl: string,
+  oauthData: DroneSecret,
+): Promise<string[]> {
+  const options: AxiosRequestConfig = {
+    params: {
+      // eslint-disable-next-line @typescript-eslint/camelcase
+      client_id: oauthData.clientId,
+      // eslint-disable-next-line @typescript-eslint/camelcase
+      redirect_uri: droneLoginUrl,
+      // eslint-disable-next-line @typescript-eslint/camelcase
+      response_type: 'code',
+    },
+    maxRedirects: 1,
+    auth: {
+      username: env.GITEA_USER,
+      password: env.GITEA_PASSWORD,
+    },
+  }
+
+  console.log('Authorizing OAuth application')
+
+  const authorizeResponse: AxiosResponse<any> = await axios.get(`${giteaUrl}/login/oauth/authorize`, options)
+
+  return authorizeResponse.headers['set-cookie']
+}
+
+export function getCookieValue(authorizeHeaderCookies: string[], cookieName: string): string {
+  // Loop over cookies and find the _csrf cookie and retrieve the value
+  const hasCookieName = (cookie: string) => cookie.includes(cookieName)
+  const cookiesWithName: string[] = authorizeHeaderCookies.filter(hasCookieName) // Find cookie that contains _csrf
+
+  if (cookiesWithName.length == 0) {
+    throw new GiteaDroneError(`No ${cookieName} cookie was returned`)
+  }
+
+  const cookieValues: string[] = cookiesWithName[0]
+    .split(';') // Split that cookie into multiple parts, as a single cookie can contain multiple key/values ';' separated
+    .map((c) => c.trim())
+    .filter(hasCookieName) // Find the key/value pair with _csrf as key
+    .map((cookie) => cookie.substring(cookieName.length + 1)) // Retrieve value for '_csrf'-key
+
+  if (cookieValues.length == 0 || cookieValues[0] === '') {
+    throw new GiteaDroneError(`No value for ${cookieName} was found`)
+  }
+  return cookieValues[0]
+}
+
+async function authorizeOAuthApp(giteaUrl: string, droneLoginUrl: string, oauthData: DroneSecret) {
+  let authorizeHeaderCookies: string[]
+  try {
+    authorizeHeaderCookies = await getGiteaAuthorizationHeaderCookies(giteaUrl, droneLoginUrl, oauthData)
+  } catch (error) {
+    throw new GiteaDroneError('Authorization already granted or something went wrong')
+  }
+
+  const csrfToken = getCookieValue(authorizeHeaderCookies, csrfCookieName)
+
+  console.log('Granting authorization')
+
+  const grantOptions: AxiosRequestConfig = {
+    method: 'POST',
+    url: `${giteaUrl}/login/oauth/grant`,
+    headers: {
+      cookie: authorizeHeaderCookies.map((cookie) => cookie.split(';')[0]).join('; '),
+      Cookie: authorizeHeaderCookies.map((cookie) => cookie.split(';')[0]).join('; '),
```

Comment:
I really want proof that this is needed to be duplicated with and without capital